address/address/address.py

```diff
@@ -0,0 +1,155 @@
+from typing import List, DefaultDict, Generic, TypeVar, Dict, NamedTuple
+import time
+import os
+import logging
+import concurrent
+import asyncio
+from aiohttp import web
+import kubernetes_asyncio as kube
+from gear import (setup_aiohttp_session, web_authenticated_developers_only,
+                  rest_authenticated_users_only, AccessLogger)
+from hailtop.config import get_deploy_config
+from hailtop.tls import get_server_ssl_context
+from web_common import setup_aiohttp_jinja2, setup_common_static_routes, render_template
+import uvloop
+import sortedcontainers
+from collections import defaultdict
+
+
+NAMESPACE = os.environ['HAIL_DEFAULT_NAMESPACE']
+
+uvloop.install()
+
+log = logging.getLogger('address')
+
+routes = web.RouteTableDef()
+
+deploy_config = get_deploy_config()
+
+
+@routes.get('/healthcheck')
+async def get_healthcheck(request):  # pylint: disable=unused-argument
+    return web.Response()
+
+
+@routes.get('/')
+@routes.get('')
+@web_authenticated_developers_only()
+async def get_index(request, userdata):  # pylint: disable=unused-argument
+    cache = request.app['cache']
+    keys = sorted(list(cache.keys))
+    page_context = {
+        'attributes': [
+            ['VALID_DURATION_IN_SECONDS', VALID_DURATION_IN_SECONDS]],
+        'rows': [{'name': k,
+                  'addresses': ", ".join([x.ip for x in cache.entries[k].value]),
+                  'ports': ", ".join([x.port for x in cache.entries[k].value]),
+                  'lifetime': time.time() - cache.entries[k].expire_time,
+                  'lock': k in cache.locks}
+                 for k in keys]
+    }
+    return await render_template('address', request, userdata, 'index.html', page_context)
+
+
+@routes.get('/api/{name}')
+@rest_authenticated_users_only
+async def get_name(request, userdata):  # pylint: disable=unused-argument
+    name = request.match_info['name']
+    log.info(f'get {name}')
+    addresses = await request.app['cache'].get(name)
+    return web.json_response([address.to_dict() for address in addresses])
+
+
+VALID_DURATION_IN_SECONDS = 60
+
+
+T = TypeVar('T')  # pylint: disable=invalid-name
+
+
+class CacheEntry(Generic[T]):
+    def __init__(self, value: T):
+        self.value: T = value
+        self.expire_time: float = time.time() + VALID_DURATION_IN_SECONDS
+
+
+class AddressAndPort(NamedTuple):
+    address: str
```

Comment:
Huh. That's an interesting format. I guess this has to do with inheriting from NamedTuple?

Reply:
yep, it's a new feature